[mlir] Fix bug that prevents application of PrintingFlags in Location

jungmair opened the pull request:
The current implementation of LocationSnapshotPass takes an OpPrintingFlags argument and stores it as member, but does not use it for printing.
Small fix: use already stored flags for printing (as originally intended)

jungmair committed:
[mlir] Fix bug that prevents application of PrintingFlags in LocationSnapshotPass

jungmair commented:
@River707 Maybe you could have a look on this?

joker-eph commented:
Thanks for the fix!
Is this something that we can exercise with a test?

jungmair commented:
> Thanks for the fix! Is this something that we can exercise with a test?

In theory yes, but IMO it is 1) highly unlikely that there will be a regression 2) requires a bit more effort because the flags are not exposed with mlir-opt and thus would require writing a C++ application that uses it with flags.

joker-eph commented:
OK that's a problem though: the flags should be made a proper pass option, can you moves these to the .td file?

jungmair commented:
I just looked into it, but it is not that easy. As far as I see, there is no previous usage of OpPrintingFlags as pass option/no existing option parser. Also OpPrintingFlags has ~9 settings, so this is not really trivial. What do you propose?

joker-eph committed:
Add support for passing selected OpPrintingFlags to the LocationSnapshot pass

joker-eph commented:
It is a strong concern to me that pass-options must roundtrip.

It does not seem really difficult though, very mechanical, see: https://github.com/joker-eph/llvm-project/commit/ceb4834144a9d221451395a04f84f984e7225cb1

More importantly, most of these are irrelevant, so I believe we can prune this down to: https://github.com/joker-eph/llvm-project/commit/94a549631f252ba324dfebd64547df2eefb9bcd5

Does this achieve what you're after?

jungmair committed:
Add test for passing OpPrintingFlags to LocationSnapshotPass in mlir-opt

jungmair commented:
Yes, thanks for your help on this.
I fixed a small typo/bug in your changes and added a test that verifies that the generated snapshot file fits the printing flags (i.e., is printed with debug info)

joker-eph commented on mlir/test/Transforms/location-snapshot.mlir:
```diff
@@ -1,5 +1,6 @@
 // RUN: mlir-opt -allow-unregistered-dialect -snapshot-op-locations='filename=%/t' -mlir-print-local-scope -mlir-print-debuginfo %s | FileCheck %s -DFILE=%/t
 // RUN: mlir-opt -allow-unregistered-dialect -snapshot-op-locations='filename=%/t tag='tagged'' -mlir-print-local-scope -mlir-print-debuginfo %s | FileCheck %s --check-prefix=TAG -DFILE=%/t
+// RUN: mlir-opt -allow-unregistered-dialect -snapshot-op-locations='filename=%/t print-debuginfo' -mlir-print-local-scope -mlir-print-debuginfo %s | FileCheck %s --check-prefix=DBG -DFILE=%/t && cat %/t | FileCheck %s --check-prefix=DBGFILE
```
Just a nit, but better avoid `$$` and split over multiple line: lit will treat it the same.
```suggestion
// RUN: mlir-opt -allow-unregistered-dialect -snapshot-op-locations='filename=%/t print-debuginfo' -mlir-print-local-scope -mlir-print-debuginfo %s | FileCheck %s --check-prefix=DBG -DFILE=%/t
// RUN: cat %/t | FileCheck %s --check-prefix=DBGFILE
```

jungmair commented:
Could you also merge it :)
